### PR TITLE
Native S2I Builder image

### PR DESCRIPTION
**Pull request by vorburger**

### CODE REVIEW

**Comment by vorburger** [comment]
How cool is this? (for #304)

@siamaksade @cescoffier @dmlloyd @gsmet @geoand @gastaldi

**Comment by geoand** [comment]
Looks great!

File: docker/centos-graal-maven-s2i/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+# This part is heavily inspired by ../centos-graal/Dockerfile
+FROM centos:latest
+
+ARG GRAAL_VERSION
+ENV GRAAL_VERSION=${GRAAL_VERSION:-1.0.0-rc12}
+ARG MAVEN_VERSION=3.6.0
+
+ENV GRAAL_CE_URL=https://github.com/oracle/graal/releases/download/vm-${GRAAL_VERSION}/graalvm-ce-${GRAAL_VERSION}-linux-amd64.tar.gz
+ENV MAVEN_URL=http://apache.osuosl.org/maven/maven-3/$MAVEN_VERSION/binaries/apache-maven-$MAVEN_VERSION-bin.tar.gz
```
Comment (gastaldi) [inline docker/centos-graal-maven-s2i/Dockerfile]:
```suggestion
ENV MAVEN_URL=https://archive.apache.org/dist/maven/maven-3/$MAVEN_VERSION/binaries/apache-maven-$MAVEN_VERSION-bin.tar.gz
```

File: docker/centos-graal-maven-s2i/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ENV MAVEN_HOME /usr/share/maven
+ENV MAVEN_CONFIG "$USER_HOME_DIR/.m2"
+ENV MAVEN_OPTS="-XX:+TieredCompilation -XX:TieredStopAtLevel=1"
+
+# This part is firmly inspired by fabric8io-images/s2i
+
+LABEL io.openshift.s2i.scripts-url="image:///usr/local/s2i" \
+      io.openshift.s2i.destination="/tmp" \
+
+      io.k8s.description="Platform for building and running Quarkus.io NATIVE Java applications" \
```
Comment (gastaldi) [inline docker/centos-graal-maven-s2i/Dockerfile]:
```suggestion
      io.k8s.description="Platform for building and running Quarkus.io Native Java applications" \
```

File: docker/centos-graal-maven-s2i/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+      io.k8s.display-name="Quarkus.io NATIVE Java Applications" \
```
Comment (gastaldi) [inline docker/centos-graal-maven-s2i/Dockerfile]:
```suggestion
      io.k8s.display-name="Quarkus.io Native Java Applications" \
```

**Review by gastaldi — APPROVED** [review]
LGTM. Added some minor suggestions, feel free to approve or reject them :smile:

**Review by gastaldi — APPROVED** [review]
LGTM

**Comment by geoand** [comment]
Very good stuff!

File: docker/centos-graal-maven-s2i/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,57 @@
+# [Quarkus.io](http://quarkus.io) *native* S2I builder
+# Original Author: Michael Vorburger.ch
+
+# This part is heavily inspired by ../centos-graal/Dockerfile
+FROM centos:latest
+
+ARG GRAAL_VERSION
+ENV GRAAL_VERSION=${GRAAL_VERSION:-1.0.0-rc12}
+ARG MAVEN_VERSION=3.6.0
```
Comment (cescoffier) [inline docker/centos-graal-maven-s2i/Dockerfile]: Wondering if we can reuse one of the docker image Quarkus provides: https://github.com/quarkusio/quarkus/tree/master/docker

Avoiding having to maintain version is several places would be great.

File: docker/centos-graal-maven-s2i/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+ENV MAVEN_HOME /usr/share/maven
+ENV MAVEN_CONFIG "$USER_HOME_DIR/.m2"
+ENV MAVEN_OPTS="-XX:+TieredCompilation -XX:TieredStopAtLevel=1"
+
+# This part is firmly inspired by fabric8io-images/s2i
+
+LABEL io.openshift.s2i.scripts-url="image:///usr/local/s2i" \
+      io.openshift.s2i.destination="/tmp" \
+
+      io.k8s.description="Platform for building and running Quarkus.io Native Java applications" \
```
Comment (cescoffier) [inline docker/centos-graal-maven-s2i/Dockerfile]: The wording is not totally in line with the website - check https://github.com/quarkusio/quarkus/blob/master/docs/src/main/asciidoc/0-glossary.adoc.

File: docker/centos-graal-maven-s2i/README.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
+# [Quarkus.io](http://quarkus.io) native S2I
+
+## How to build this S2I builder image
+
+### OpenShift
+
+    oc new-build https://github.com/quarkusio/quarkus.git --context-dir=docker/centos-graal-maven-s2i --name quarkus-native-s2i
```
Comment (cescoffier) [inline docker/centos-graal-maven-s2i/README.md]: Nice, just learned something :-D

File: docker/centos-graal-maven-s2i/README.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+### Locally
+
+    docker build . -t quarkus-native-s2i
+
+## How to use this
+
+### OpenShift
+
+    oc new-app quarkus-native-s2i~https://github.com/quarkusio/quarkus-quickstarts --context-dir=getting-started-native --name=getting-started-native
```
Comment (cescoffier) [inline docker/centos-graal-maven-s2i/README.md]: Should we remind the user to have "installed" with the command mentioned previously?

File: docker/centos-graal-maven-s2i/README.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+Note that GraalVM-based native build are more memory & CPU intensive than regular pure Java builds,
+and you therefore may need to increase the quota for OpenShift's S2I builder containers.
```
Comment (cescoffier) [inline docker/centos-graal-maven-s2i/README.md]: Can you give some hint about this? 
I guess you can edit the build object.

File: docker/centos-graal-maven-s2i/s2i/assemble [hunk 9]
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh
+
+# This POC script is based on the openshift/source-to-image project documentation,
+# and loosely inspired by fabric8io-images/s2i's assemble script, but **MUCH** simplified;
+# a TODO future PRODUCTION version of this would probably want to re-use that script...
+
+set -ex
+
+cd /tmp/src/
+mvn package -Pnative -DskipTests -Dmaven.javadoc.skip=true -Dmaven.site.skip=true -Dmaven.source.skip=true -Djacoco.skip=true -Dcheckstyle.skip=true -Dfindbugs.skip=true -Dpmd.skip=true -Dfabric8.skip=true -e -B
```
Comment (cescoffier) [inline docker/centos-graal-maven-s2i/s2i/assemble]: That's a lot of skip :-) But I'm all good with that.

Comment (cescoffier) [reply docker/centos-graal-maven-s2i/s2i/assemble]: Maybe you can use multiline as the `-e -B` are quite important (and hidden right now)

File: docker/centos-graal-maven-s2i/s2i/run [hunk 10]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+
+# This POC script is (intentionally) overly trivial.
+# a TODO future PRODUCTION version of this would probably want tomake it possible to
+# e.g. pass additional parameters to the executable, as does fabric8io-images/s2i's run script.
+
+/quarkus.io/*-runner -Dquarkus.http.host=0.0.0.0
```
Comment (cescoffier) [inline docker/centos-graal-maven-s2i/s2i/run]: For the record, there are discussions about the 0.0.0.0 thingy.

**Review by cescoffier — CHANGES_REQUESTED** [review]
That's pretty cool, I've made a few comments.

**Comment by cescoffier** [comment]
BTW, how will we version the S2I, during the release process? If so how does that work?

File: docker/centos-graal-maven-s2i/Dockerfile (hunk 4 above)
Comment (vorburger) [reply docker/centos-graal-maven-s2i/Dockerfile]: @cescoffier saw those and took inspiration from them.. the problem with doing e.g. a `FROM centos-graal-native-image` is that it makes it a bit harder to then set up CI/CD of this container - systems like Quay and Docker Hub except to run (exactly) 1 Dockerfile... so, at least for a first iteration, it seems simpler to me to keep it all together.

File: docker/centos-graal-maven-s2i/Dockerfile (hunk 5 above)
Comment (vorburger) [reply docker/centos-graal-maven-s2i/Dockerfile]: I've attempted to adhere to that in an upcoming update (and raised #1377), hopefully better now.

File: docker/centos-graal-maven-s2i/README.md (hunk 7 above)
Comment (vorburger) [reply docker/centos-graal-maven-s2i/README.md]: I'll re-shuffle the documents sections to make more sense, see if you like it better after the next update.

File: docker/centos-graal-maven-s2i/README.md (hunk 8 above)
Comment (vorburger) [reply docker/centos-graal-maven-s2i/README.md]: I was hoping someone else would help with how exactly to do that... :smiley:

File: docker/centos-graal-maven-s2i/s2i/assemble (hunk 9 above)
Comment (vorburger) [reply docker/centos-graal-maven-s2i/s2i/assemble]: FYI The skip are [are stolen from here](https://github.com/fabric8io-images/s2i/blob/master/java/templates/s2i/assemble#L123). I've moved `-e -B` to the front and made it multi line.

File: docker/centos-graal-maven-s2i/s2i/run (hunk 10 above)
Comment (vorburger) [reply docker/centos-graal-maven-s2i/s2i/run]: Yup, so I gathered.. if/when that happens, someone needs to drop this from a whole lot of quickstarts and docs and scripts and all - including this. For now, it's required though.

**Comment by vorburger** [comment]
> BTW, how will we version the S2I, during the release process? If so how does that work?

Not 100% sure I know what you mean, but if you are asking how we will release a container image from this Dockerfile to e.g. Quay.io (and/or Docker Hub), then that's trivial to set up (IFF you keep a single `Dockerfile` ... see above), and it can easily be made to "tag" such images based on the branch in this repo. Does this help?

File: docker/centos-graal-maven-s2i/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,57 @@
+# [Quarkus.io](http://quarkus.io) *native* S2I builder
+# Original Author: Michael Vorburger.ch
+
+# This part is heavily inspired by ../centos-graal/Dockerfile
+FROM centos:latest
+
+ARG GRAAL_VERSION
+ENV GRAAL_VERSION=${GRAAL_VERSION:-1.0.0-rc12}
```
Comment (cescoffier) [inline docker/centos-graal-maven-s2i/Dockerfile]: Need to be updated to 13.

File: docker/centos-graal-maven-s2i/Dockerfile (hunk 4 above)
Comment (cescoffier) [reply docker/centos-graal-maven-s2i/Dockerfile]: Ok

File: docker/centos-graal-maven-s2i/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,57 @@
+# [Quarkus.io](http://quarkus.io) *native* S2I builder
+# Original Author: Michael Vorburger.ch
+
+# This part is heavily inspired by ../centos-graal/Dockerfile
+FROM centos:latest
```
Comment (cescoffier) [inline docker/centos-graal-maven-s2i/Dockerfile]: Do we really want to use "latest"?

File: docker/centos-graal-maven-s2i/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+      io.openshift.tags="builder,java,quarkus,native"
+
+# S2I scripts + README
+COPY s2i /usr/local/s2i
+RUN chmod 755 /usr/local/s2i/*
+ADD README.md /usr/local/s2i/usage.txt
+
+ENV PATH=$PATH:"/usr/local/s2i"
+
+RUN mkdir /quarkus.io/
```
Comment (cescoffier) [inline docker/centos-graal-maven-s2i/Dockerfile]: Really minor, but I try to about `.` in my directory name. What about just "quarkus"

File: docker/centos-graal-maven-s2i/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+EXPOSE 8080
```
Comment (cescoffier) [inline docker/centos-graal-maven-s2i/Dockerfile]: Any reason to expose the port?

**Review by cescoffier — CHANGES_REQUESTED** [review]
A few comments

File: docker/centos-graal-maven-s2i/Dockerfile (hunk 12 above)
Comment (vorburger) [reply docker/centos-graal-maven-s2i/Dockerfile]: Using e.g. `centos:7` is actually meaningless and doesn't work the way you think it does, as far as I understand... I learnt this the hard way over here https://github.com/fabric8io-images/s2i/issues/206#issuecomment-445047444 and here https://github.com/fabric8io-images/s2i/issues/172 full background logged here https://gist.github.com/vorburger/ea367a76f69aafb8260c62f6bedf2788... Unless all that's only for 7.5/7.6 and not 8, but I doubt it. I am still happy to change it to `centos:7` anyway if you like, but doubt that's really useful and does what most expect.

File: docker/centos-graal-maven-s2i/Dockerfile (hunk 14 above)
Comment (vorburger) [reply docker/centos-graal-maven-s2i/Dockerfile]: Just makes it easier to use `oc new-app` / `oc expose`  (and `docker run`) without having to re-specify `8080` ... of course one could change it, but ... people probably typically will not? And even if they did (and therefore specified another port in `oc new-app` / `oc expose`  (and `docker run`), this doesn't cause any "harm" either, for the convenience it offers? (See also https://github.com/fabric8io-images/s2i/issues/115 and https://github.com/fabric8io-images/s2i/pull/205 for some history.)

File: docker/centos-graal-maven-s2i/README.md (hunk 8 above)
Comment (vorburger) [reply docker/centos-graal-maven-s2i/README.md]: There now is at least some doc. Could always be further improved, but god enough for a 0.1?

File: docker/centos-graal-maven-s2i/Dockerfile (hunk 11 above)
Comment (vorburger) [reply docker/centos-graal-maven-s2i/Dockerfile]: done! Thanks for noticing.

File: docker/centos-graal-maven-s2i/Dockerfile (hunk 13 above)
Comment (vorburger) [reply docker/centos-graal-maven-s2i/Dockerfile]: it's now `/home/quarkus` OK for you?

**Comment by vorburger** [comment]
@cescoffier I've further tested this, and (from my side) this is good for merge now.

File: docker/centos-graal-maven-s2i/Dockerfile (hunk 13 above)
Comment (cescoffier) [reply docker/centos-graal-maven-s2i/Dockerfile]: Yes

File: docker/centos-graal-maven-s2i/Dockerfile (hunk 14 above)
Comment (cescoffier) [reply docker/centos-graal-maven-s2i/Dockerfile]: Thanks for the explanation.

**Review by cescoffier — APPROVED** [review]

**Comment by cescoffier** [comment]
Waiting for CI and will merge.

File: docker/centos-graal-maven-s2i/s2i/run [hunk 15]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+
+# This POC script is (intentionally) overly trivial.
+# a TODO future PRODUCTION version of this would probably want tomake it possible to
+# e.g. pass additional parameters to the executable, as does fabric8io-images/s2i's run script.
+
+${QUARKUS_HOME}/*-runner -Dquarkus.http.host=0.0.0.0
```
Comment (geoand) [inline docker/centos-graal-maven-s2i/s2i/run]: `-Dquarkus.http.host=0.0.0.0` shouldn't be needed anymore since it's now the default

Comment (vorburger) [reply docker/centos-graal-maven-s2i/s2i/run]: But is that already released, right now? I would prefer to keep this and then see it removed a while after a new release, together with more [generally cleaning this up in many places where it currently still appears](https://github.com/search?q=org%3Aquarkusio+quarkus.http.host&type=Code).

File: docker/centos-graal-maven-s2i/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,61 @@
+# [Quarkus.io](http://quarkus.io) *native* S2I builder
+# Original Author: Michael Vorburger.ch
+
+# This part is heavily inspired by ../centos-graal/Dockerfile
+FROM centos:latest
+
+ARG GRAAL_VERSION=1.0.0-rc13
```
[the remark this answers is not in the document]
Comment (vorburger) [reply docker/centos-graal-maven-s2i/Dockerfile]: Yup, just like in [a few other places where it appears](https://github.com/search?q=org%3Aquarkusio+1.0.0-rc13&type=Code).

File: docker/centos-graal-maven-s2i/s2i/run (hunk 15 above)
Comment (geoand) [reply docker/centos-graal-maven-s2i/s2i/run]: The fix is in master and will be present in the `0.12.0` release that I think @cescoffier will be doing tomorrow. 
I'll leave it up to you to best decide how to handle this :)

File: docker/centos-graal-maven-s2i/Dockerfile (hunk 16 above)
[the remark this answers is not in the document]
Comment (geoand) [reply docker/centos-graal-maven-s2i/Dockerfile]: :+1

File: docker/centos-graal-maven-s2i/s2i/run (hunk 15 above)
Comment (geoand) [reply docker/centos-graal-maven-s2i/s2i/run]: https://github.com/quarkusio/quarkus/pull/1312#issuecomment-474271302

**Comment by vorburger** [comment]
> Waiting for CI and will merge.

@cescoffier merge?

**Comment by cescoffier** [comment]
Thanks!

**Pull request merged**